File: clients/python/baml_lib/_impl/deserializer/raw_wrapper/raw_wrapper.py
import abc
import typing


class RawWrapper(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def as_str(self, inner: bool) -> typing.Optional[str]:
        raise NotImplementedError()

    # Try to return a proper string representation. For example this would strip quotes
    # or if turning a list into a str, it may just return the 1st item if there is only one without
    # the [].
    @abc.abstractmethod
    def as_smart_str(self, inner: bool) -> typing.Optional[str]:
        raise NotImplementedError()

    @abc.abstractmethod
    def as_self(self) -> typing.Any:
        raise NotImplementedError()

    @abc.abstractmethod
    def as_int(self) -> typing.Optional[int]:
        raise NotImplementedError()

    @abc.abstractmethod
    def as_float(self) -> typing.Optional[float]:
        raise NotImplementedError()

    @abc.abstractmethod
    def as_bool(self) -> typing.Optional[bool]:
        raise NotImplementedError()

    @abc.abstractmethod
    def as_list(self) -> typing.Iterable["RawWrapper"]:
        raise NotImplementedError()

    @abc.abstractmethod
    def as_dict(
        self,
    ) -> typing.ItemsView[typing.Optional["RawWrapper"], "RawWrapper"]:
        raise NotImplementedError()

File: clients/ts/src/baml_lib/deserializer/raw_wrapper/raw_wrapper.ts
// TypeScript equivalent of Python's 'abc' and 'typing' modules
abstract class RawWrapper {
    // Abstract method as_str
    abstract as_str(inner: boolean): string | undefined;

    // Abstract method as_smart_str
    abstract as_smart_str(inner: boolean): string | undefined;

    // Abstract method as_self
    abstract as_self(): any;

    // Abstract method as_int
    abstract as_int(): number | undefined;

    // Abstract method as_float
    abstract as_float(): number | undefined;

    // Abstract method as_bool
    abstract as_bool(): boolean | undefined;

    // Abstract method as_list
    abstract as_list(): Iterable<RawWrapper>;

    // Abstract method as_dict
    abstract as_dict(): Iterable<[RawWrapper | undefined, RawWrapper]>;
}

export { RawWrapper };
